db/blob/blob_source.cc

```diff
@@ -0,0 +1,192 @@
+//  Copyright (c) Meta Platforms, Inc. and affiliates.
+//  This source code is licensed under both the GPLv2 (found in the
+//  COPYING file in the root directory) and Apache 2.0 License
+//  (found in the LICENSE.Apache file in the root directory).
+
+#include "db/blob/blob_source.h"
+
+#include <cassert>
+#include <string>
+
+#include "db/blob/blob_file_reader.h"
+#include "file/file_prefetch_buffer.h"
+#include "options/cf_options.h"
+
+namespace ROCKSDB_NAMESPACE {
+
+BlobSource::BlobSource(Cache* cache, const ImmutableOptions* immutable_options,
+                       const FileOptions* file_options,
+                       const std::string& db_id,
+                       const std::string& db_session_id,
+                       uint32_t column_family_id,
+                       HistogramImpl* blob_file_read_hist,
+                       const std::shared_ptr<IOTracer>& io_tracer)
+    : db_id_(db_id),
+      db_session_id_(db_session_id),
+      statistics_(immutable_options->statistics.get()),
+      blob_file_cache_(new BlobFileCache(cache, immutable_options, file_options,
+                                         column_family_id, blob_file_read_hist,
+                                         io_tracer)),
+      blob_cache_(immutable_options->blob_cache) {}
+
+BlobSource::~BlobSource() = default;
+
+Status BlobSource::GetBlobFromCache(const Slice& cache_key, Cache* blob_cache,
+                                    CachableEntry<PinnableSlice>* blob) const {
+  assert(blob);
+  assert(blob->IsEmpty());
+
+  if (blob_cache != nullptr) {
+    assert(!cache_key.empty());
+    Cache::Handle* cache_handle = nullptr;
+    cache_handle = GetEntryFromCache(blob_cache, cache_key);
+    if (cache_handle != nullptr) {
+      blob->SetCachedValue(
+          static_cast<PinnableSlice*>(blob_cache->Value(cache_handle)),
+          blob_cache, cache_handle);
+      return Status::OK();
+    }
+  }
+
+  assert(blob->IsEmpty());
+
+  return Status::NotFound("Blob record not found in cache");
+}
+
+Status BlobSource::PutBlobIntoCache(const Slice& cache_key, Cache* blob_cache,
+                                    CachableEntry<PinnableSlice>* cached_blob,
+                                    PinnableSlice* blob) const {
+  assert(blob);
+  assert(!cache_key.empty());
+
+  const Cache::Priority priority = Cache::Priority::LOW;
+
+  Status s;
+
+  if (blob_cache != nullptr) {
+    // Objects to be put into the cache have to be heap-allocated and
+    // self-contained, i.e. own their contents. The Cache has to be able to take
+    // unique ownership of them. Therefore, we copy the blob into a separate
+    // heap-allocated PinnableSlice, and insert that into the cache.
+    PinnableSlice* value = new PinnableSlice();
```

Comment:
Any reason not to use std::string directly (save ~25 bytes per entry)?

For future consideration: a bit of hacking could remove a level of indirection and use a LengthPrefixedSlice in a char[]

Reply:
Yeah, good point re: `std::string`

Reply:
done